File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def reverse_linked_list(start):
            prev = None
            curr = start

            while curr:
                next_node = curr.next
                curr.next = prev
                prev = curr
                curr = next_node

            return prev

        dummy = ListNode()
        dummy.next = head
        prev_group_end = dummy

        while True:
            # find the start and the end of the group
            group_start = prev_group_end.next
            group_end = prev_group_end

            for _ in range(k):
                group_end = group_end.next
                if not group_end:
                    return dummy.next

            next_group_start = group_end.next

            group_end.next = None
            new_group_start = reverse_linked_list(group_start)

            prev_group_end.next = new_group_start
            group_start.next = next_group_start

            prev_group_end = group_start

File: Data Structures & Algorithms/reorder-linked-list/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        // find middle
        ListNode* slow = head;
        ListNode* fast = head;

        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // break it
        ListNode* prev = nullptr;
        ListNode* curr = slow->next;
        slow->next = nullptr;

        // reverse the second half
        while (curr != nullptr) {
            ListNode* next_node = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next_node;
        }

        // weave
        ListNode* first = head;
        ListNode* second = prev;

        while (second != nullptr) {
            ListNode* temp1 = first->next;
            ListNode* temp2 = second->next;
            
            first->next = second;
            second->next = temp1;

            first = temp1;
            second = temp2;
        }
    }
};
